script/model/model_server.py

```diff
@@ -0,0 +1,355 @@
+        :param result: Any result
+        :return:
+        """
+        return {
+            "action": action,
+            "result": result
+        }
+
+    @staticmethod
+    def _parse_msg(payload:str) -> Tuple[int, int, Message]:
+        logging.debug("PY RECV: " + payload)
+        tokens = payload.split('-', 2)
+
+        # Invalid message format
+        try:
+            msg_id = int(tokens[0])
+            recv_id = int(tokens[1])
+        except ValueError as e:
+            logging.error(f"Invalid message payload format: {payload}, ids not int.")
+            return -1, -1, None
+
+        msg = Message.from_json(tokens[2])
+        return msg_id, recv_id, msg
+
+    @staticmethod
+    def _transform_opunit_data(raw_data: Dict) -> List[opunit_data.OpUnitData]:
+        """
+        Transform the raw training data from ModelServerManger for each data unit
+        :param raw_data: {
+            opunit: (x, y)
+        }
+        :return:
+        """
+        # TODO(ricky): serialize the opunit by name rather than index
+        return [opunit_data.OpUnitData(OpUnit(x[0]), x[1][0], x[1][1]) for x in raw_data]
+
+    def _train_model(self, data: Dict) -> str:
+        """
+        Train a model with the given model name and seq_files directory
+        :param data: {
+            models: [lr, XXX, ...],
+            seq_files: PATH_TO_SEQ_FILES_FOLDER, or None
+            data_lists: [(OpUnit, X_List, Y_List)/OpUnitData]
+        }
+        :return: the model map
+        """
+        ml_models = data["models"]
+        seq_files_dir = data["seq_files"]
+        raw_data = data["raw_data"]
+
+        result_path = ModelServer.RESULT_PATH
+        test_ratio = ModelServer.TEST_RATIO
+        trim = ModelServer.TRIM_RATIO
+        expose_all = ModelServer.EXPOSE_ALL
+        txn_sample_interval = ModelServer.TXN_SAMPLE_INTERVAL
+        trainer = MiniTrainer(seq_files_dir, result_path, ml_models, test_ratio, trim, expose_all,txn_sample_interval)
+        if seq_files_dir:
+            # Perform training from MiniTrainer and input files directory
+            model_map = trainer.train()
+        else:
+            """
+            NOTE:
+            This branch has not been tested yet because the MiniTrainer now heavily relies on the CSV file. 
+            So training with data directly is not yet possible without refactoring the MiniTrainer 
+            """
+
+            # Train from data directly
+            assert(raw_data)
+
+            # Transform with raw data lists
+            opunit_data_list = self._transform_opunit_data(raw_data)
+            for data in opunit_data_list:
+                best_y_transformer, best_method = trainer.train_data(data, None)
+                if trainer.expose_all:
+                    trainer.train_specific_model(data, best_y_transformer, best_method)
+
+            # Directly get the model map
+            model_map = trainer.get_model_map()
+
+        self.g_model_map = model_map
```

Comment:
This would make the `model_server` stateful. I think we can remove this line with the above suggested refactor.

Reply:
I think one question I have is: would there be multiple models trained at the same time?

Reply:
There can be multiple models trained at the same time, but they would be in the same model map.